hail/src/main/resources/include/hail/Region2.h

```diff
@@ -0,0 +1,72 @@
+#ifndef HAIL_REGION2_H
+#define HAIL_REGION2_H 1
+
+#include <memory>
+#include <vector>
+#include <utility>
+#include "hail/Upcalls.h"
+#include "hail/NativeStatus.h"
+
+namespace hail {
+
+class RegionPool {
+  static constexpr ssize_t block_size = 64*1024;
+  static constexpr ssize_t block_threshold = 4096;
+
+  public:
+    class Region {
+      private:
+        RegionPool * pool_;
+        ssize_t block_offset_;
+        char * current_block_;
+        std::vector<char *> used_blocks_;
+        std::vector<char *> big_chunks_;
+        std::vector<std::shared_ptr<Region>> parents_;
+        char * allocate_new_block();
+        char * allocate_big_chunk(ssize_t size);
+      public:
+        Region(RegionPool * pool);
+        void clear();
+        inline char * allocate(ssize_t alignment, ssize_t n) {
+          ssize_t aligned_off = (block_offset_ + alignment - 1) & ~(alignment - 1);
+          if (aligned_off + n <= block_size) {
+            char* p = current_block_ + aligned_off;
+            block_offset_ = aligned_off + n;
+            return p;
+          } else {
+            return (n <= block_threshold) ? allocate_new_block() : allocate_big_chunk(n);
+          }
+        }
+        std::shared_ptr<Region> get_region();
+        void add_reference_to(std::shared_ptr<Region> region);
+    };
+
+  private:
+    std::vector<Region *> free_regions_;
```

Comment:
How much benefit do you think we get from having a pool of empty regions? I would imagine the pool of chunks is mostly what matters. The region pool has some tricky ownership semantics, but without it you would only need `Region2` to have a destructor that calls `clear()` to return its chunks to the pool before dying. It also means every new region can be created by `make_shared`, which puts the reference count in the same allocation as the region object, which means all reference count adjustments will have better cache locality.

Reply:
I'm not super sure on this point. @cseed? I will say that I don't anticipate reference counting to get super involved---I don't really see a scenario where we'll have more than 2 references to a given region.

Reply:
By tricky ownership semantics, I was referring to how a region goes back and forth between being owned by a collection of shared pointers and being owned (implicitly) by the `RegionPool`. That's not so bad, but the only allocations it saves are those backing the vectors of pointers to used blocks, which should be very small, and are only made O(num rows) times. As I mentioned keeping the region pool does still require a small allocation each time it's taken out of the pool for the reference count, so it's not clear it saves much.

Not a big deal, but it strikes me as a premature optimization. I'm fine with whichever way you want to go.

Reply:
yeah, that's the part I wasn't super sure on either. It's definitely something I'd like to benchmark and compare as we start actually using it. I could run some tests now though, if you'd like.

Reply:
Up to you.

Reply:
I added a benchmark test to this PR and ran it on both the current version and on a version with: 

```
std::shared_ptr<Region2> RegionPool::get_region() {
  return std::make_shared<Region2>(this);
}
```

The current version, where the region pool manages the allocation and unused regions get returned to the pool instead of freed, appears to be a bit faster, especially in what I think will be the expected use case where we end up needing a small(er) number of regions at a time, but free and acquire new regions to process each row.